### PR TITLE
Handling teststep screen references missing 

### PR DESCRIPTION
**Pull request by NavneetThekkumpat**
Problem: Controls can change screens, screens can be renamed, but teststep value isn't updated and causes errors while roudntripping.

Solution: Convert the errors to warning and using entropy to store the screen id without screens.

### CODE REVIEW

File: src/PAModelTests/ReadTransformTests.cs [hunk 1]
```diff
@@ -24,5 +24,22 @@ public void AfterReadNullChildren(string filename)
             msapp.ApplyAfterMsAppLoadTransforms(errorContainer);
             Assert.IsFalse(errorContainer.HasErrors);
         }
+
+        [DataTestMethod]
+        [DataRow("TestStepWithInvalidScreen.msapp")]
+        public void TestStepWithInvalidScreenAfterRead(string filename)
```
Comment (jt000) [inline src/PAModelTests/ReadTransformTests.cs:30]: Could probably make this more re-usable by introducing multiple parameters (filename, has errors, has warnings) which could share the test code with `AfterReadNullChildren` and possibly more...
```
        [DataTestMethod]
        [DataRow("GalleryTemplateNullChildren.msapp", false, false)]
        [DataRow("TestStepWithInvalidScreen.msapp", false, true)]
        public void ApplyAfterMsAppLoadTransforms_Test(string filename, bool hasErrors, bool hasWarnings)
        {
            var path = Path.Combine(Environment.CurrentDirectory, "Apps", filename);
            Assert.IsTrue(File.Exists(path));
            (var msapp, var errorContainer) = CanvasDocument.LoadFromMsapp(path);
            errorContainer.ThrowOnErrors();
            msapp.ApplyAfterMsAppLoadTransforms(errorContainer);
            Assert.AreEqual(errorContainer.HasErrors, hasErrors);
            Assert.AreEqual(errorContainer.HasWarnings, hasWarnings);
        }
```

File: src/PAModel/SourceTransforms/AppTestTransform.cs [hunk 2]
```diff
@@ -105,16 +105,16 @@ public void AfterRead(BlockNode control)
                 {
                     if (!_screenIdToScreenName.ToDictionary(kvp => kvp.Key, kvp => kvp.Value).TryGetValue(testStep.ScreenId, out var screenName))
                     {
-                        _errors.ValidationError($"ScreenId referenced by TestStep {testStep.Rule} in {control.Name.Identifier} could not be found");
-                        throw new DocumentException();
+                        _errors.ValidationWarning($"ScreenId referenced by TestStep {testStep.Rule} in {control.Name.Identifier} could not be found");
+                        _entropy.RuleScreenIdWithoutScreen[testStep.Rule] = testStep.ScreenId;
```
[the remark this answers is not in the document]
Comment (NavneetThekkumpat) [reply src/PAModel/SourceTransforms/AppTestTransform.cs:109]: That makes sense and yes it looks like it is possible. Thank you for pointing this out. Looking into another possible unique key.

**Review by landonarogers — APPROVED** [review]

File: src/PAModel/SourceTransforms/AppTestTransform.cs [hunk 3]
```diff
@@ -105,16 +105,26 @@ public void AfterRead(BlockNode control)
...
-                        _errors.ValidationError($"ScreenId referenced by TestStep {testStep.Rule} in {control.Name.Identifier} could not be found");
-                        throw new DocumentException();
+                        _errors.ValidationWarning($"ScreenId referenced by TestStep {testStep.Rule} in {control.Name.Identifier} could not be found");
+                        var testStepRuleKey = $"{control.Name.Identifier}.{testStep.Rule}";
+
+                        // checking if this key already exist, not an ideal situation
+                        // Fallback logic just to make sure collisions are avoided.
+                        if (_entropy.RuleScreenIdWithoutScreen.ContainsKey(testStepRuleKey))
+                        {
+                            _errors.ValidationWarning($"RuleScreenIdWithoutScreen has a duplicate key {testStepRuleKey}");
```
Comment (jt000) [inline src/PAModel/SourceTransforms/AppTestTransform.cs:115]: If we're throwing an exception, should this remain an Error?

File: src/PAModelTests/ReadTransformTests.cs [hunk 4]
```diff
@@ -11,18 +11,21 @@ namespace PAModelTests
 {
     [TestClass]
     public class ReadTransformTests
-    {   
+    {  
         [DataTestMethod]
-        [DataRow("GalleryTemplateNullChildren.msapp")]
-        public void AfterReadNullChildren(string filename)
+        [DataRow("GalleryTemplateNullChildren.msapp", false, false)]
+        [DataRow("TestStepWithInvalidScreen.msapp", false, true)]
```
Comment (jt000) [inline src/PAModelTests/ReadTransformTests.cs:17]: Is this a test for the RuleScreenIdWithoutScreen warning\error?

Comment (NavneetThekkumpat) [reply src/PAModelTests/ReadTransformTests.cs:17]: Yes, a test .msapp that reproduced this error and is fixed now.

File: src/PAModel/SourceTransforms/AppTestTransform.cs (hunk 3 above)
Comment (NavneetThekkumpat) [reply src/PAModel/SourceTransforms/AppTestTransform.cs:115]: Yup makes sense, I can change it to an error and remove the names.

**Pull request merged**